extensions/amp-analytics/0.1/visibility-helper.js

```diff
@@ -382,15 +409,464 @@ export class VisibilityHelper {
+
+      // Optionally, element-level state.
+      const resource = opt_element ?
+          this.resources_.getResourceForElementOptional(opt_element) : null;
+      if (resource) {
+        const layoutBox = resource.getLayoutBox();
+        Object.assign(state, {
+          'elementX': layoutBox.left,
+          'elementY': layoutBox.top,
+          'elementWidth': layoutBox.width,
+          'elementHeight': layoutBox.height,
+        });
+      }
+
+      callback(state);
+    });
+
+    this.models_.push(model);
+    model.unsubscribe(() => {
+      const index = this.models_.indexOf(model);
+      if (index != -1) {
+        this.models_.splice(index, 1);
+      }
+    });
+
+    // Observe the element via InOb.
+    if (opt_element) {
+      // It's important that this happens after all the setup is done, b/c
+      // intersection observer can fire immedidately. Per spec, this should
+      // NOT happen. However, all of the existing InOb polyfills, as well as
+      // some versions of native implementations, make this mistake.
+      model.unsubscribe(this.observe(opt_element, () => model.update()));
+    }
+
+    // Start update.
+    model.update();
+    return function() {
+      model.dispose();
+    };
+  }
+
+  /**
+   * Observes the intersections of the specified element in the viewport.
+   * @param {!Element} element
+   * @param {function(number)} unusedListener
+   * @return {!UnlistenDef}
+   * @protected
+   * @abstract
+   */
+  observe(element, unusedListener) {}
+
+  /**
+   * @param {!Element} element
+   * @return {number}
+   * @abstract
+   */
+  getElementVisibility(element) {}
 }
 
 
 /**
- * Calculates the specified time based on the given `baseTime`.
- * @param {time} time
- * @param {time} baseTime
- * @return {time}
+ * The implementation of `VisibilityManager` for an AMP document. Two
+ * distinct modes are supported: the main AMP doc and a in-a-box doc.
  */
-function timeBase(time, baseTime) {
-  return time >= baseTime ? time - baseTime : 0;
+export class VisibilityManagerForDoc extends VisibilityManager {
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    super(/* parent */ null, ampdoc);
+
+    /** @const @private */
+    this.viewer_ = viewerForDoc(ampdoc);
+
+    /** @const @private */
+    this.viewport_ = viewportForDoc(ampdoc);
+
+    /** @private {boolean} */
+    this.backgrounded_ = !this.viewer_.isVisible();
+
+    /** @const @private {boolean} */
+    this.backgroundedAtStart_ = this.isBackgrounded();
+
+    /**
+     * @const
+     * @private {!Object<number, {
+     *   element: !Element,
+     *   intersectionRatio: number,
+     *   listeners: !Array<function(number)>
+     * }>}
+     */
+    this.trackedElements_ = map();
+
+    /** @private {?IntersectionObserver|?IntersectionObserverPolyfill} */
+    this.intersectionObserver_ = null;
+
+    if (getMode(this.ampdoc.win).runtime == 'inabox') {
```

Comment:
It's already lazy-initialized. The VisibilityManager is not created until the first visibility trigger is requested.